op-chain-ops/script/deploy.go

```diff
@@ -0,0 +1,330 @@
+package script
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/ethereum-optimism/optimism/op-chain-ops/foundry"
+	"github.com/ethereum-optimism/optimism/op-chain-ops/script/addresses"
+	"github.com/holiman/uint256"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// ForgeScriptBackend is a minimal interface suitable for deploying & interacting with
+// foundry scripts on chain
+type ForgeScriptBackend interface {
+	// Call sends a transaction with a []byte data to a specified address
+	Call(to common.Address, input []byte) (result []byte, err error)
+
+	// Deploy deploys a contract based on an artifact, attaches metadata and labels the deployment
+	Deploy(artifact *foundry.Artifact, label string) (address common.Address, err error)
+
+	// Destroy wipes the specified address
+	Destroy(address common.Address)
+}
+
+// ForgeScript is a generic script instance
+type ForgeScript interface {
+	// Underlying script ABI
+	ABI() abi.ABI
+
+	// Script name (mostly for logging purposes)
+	Name() string
+
+	// Sends the input as a payload to the script contract
+	Call(input []byte) (result []byte, err error)
+}
+
+// DeployScriptWithoutOutput is a specific ForgeScript that accepts typed input
+// and calls a specific script entrypoint (the run method by convention)
+//
+// The method is assumed to return nothing (empty bytes) and the output is discarded
+type DeployScriptWithoutOutput[I any] interface {
+	ForgeScript
+	Run(input I) (err error)
+}
+
+// DeployScriptWithOutput is a specific ForgeScript that accepts typed input
+// and calls a specific script entrypoint (the run method by convention)
+//
+// The method is assumed to return a single value of type O
+type DeployScriptWithOutput[I any, O any] interface {
+	ForgeScript
+	Run(input I) (output O, err error)
+}
+
+// We make sure that our implementations match the interfaces above
+var (
+	_ ForgeScriptBackend               = (*forgeScriptBackendImpl)(nil)
+	_ ForgeScript                      = (*forgeScriptImpl)(nil)
+	_ DeployScriptWithoutOutput[any]   = (*deployScriptWithoutOutputImpl[any])(nil)
+	_ DeployScriptWithOutput[any, any] = (*deployScriptWithOutputImpl[any, any])(nil)
+)
+
+// NewForgeScriptBackend creates an instance of ForgeScriptBackend
+func NewForgeScriptBackend(host *Host) ForgeScriptBackend {
+	return &forgeScriptBackendImpl{
+		host: host,
+	}
+}
+
+// NewForgeScriptFromFile loads a script artifact from the artifact filesystem on the host,
+// creates a ForgeScriptBackend from the host and finally creates a ForgeScript instance based on these
+func NewForgeScriptFromFile(host *Host, fileName string, name string) (ForgeScript, error) {
+	artifact, err := host.Artifacts().ReadArtifact(fileName, name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load script %s from %s: %w", name, fileName, err)
+	}
+
+	backend := NewForgeScriptBackend(host)
```

Comment:
Why the extra layer of indirection here?

Reply:
To decouple interface from implementation. Also it allows us to test `ForgeScriptBackend` and `ForgeScript` separately:

- `ForgeScriptBackend` has a complex`Deploy` that can go wrong or right in many places
- `ForgeScript` has a complex `Call` that can go wrong or right in many places

If we didn't decouple, we would need to be testing the deploy process everytime we're trying to test the call process. Once we decouple, we just test the deploy process separately and mock the success/failure for testing the call process.